architecture_records/002-acceleration-framework.md

```diff
@@ -0,0 +1,302 @@
+All proposed code changes to `sft_trainer.py` contained in minimal lines of code:
+- Plugins loaded by discovery; transparent to `sft_trainer.py`.
+- Plugin configuration automatically parsed.
+- Passthrough to original operation if `Framework` is disabled.
+
+```python
+from tuning.acceleration import AccelerationFramework
+
+# Minor Change 1: creating the framework object
+framework = None
+if framework_args.config_file is not None:
+    framework = AccelerationFramework(framework_args.config_file)
+
+# Minor Change 2: custom loader (if necessary)
+_model_loader = AutoModelForCausalLM.from_pretrained # default
+if framework is not None and framework.requires_custom_loading:
+    _model_loader = framework.model_loader # drop in replacement
+
+# will passthrough the default loader if framework is disabled
+model = _model_loader(
+    model_args.model_name_or_path,
+    cache_dir=train_args.cache_dir,
+    torch_dtype=get_torch_dtype(model_args.torch_dtype),
+    attn_implementation="flash_attention_2" if model_args.use_flash_attn else None,
+)
+
+# Minor Change 3: 
+if framework is not None and framework.requires_agumentation:
+    # will also take in some other configs that may affect augmentation
+    # some of these args may be modified due to the augmentation
+    # e.g., peft_config will be consumed in augmentation, and returned as None 
+    #       to prevent SFTTrainer from doing extraneous PEFT logic
+    model, (peft_config,) = framework.augmentation(
+        model, trainer.accelerator,
+        train_args, modifiable_args=(peft_config,),
+    )
+
+# instantiate trainer. Pass in model (with training enchancements)
+trainer = Trainer(model, ...)
+
+# Minor Change 4: add trainer callbacsk
+trainer.add_callbacks(framework.callbacks())
+
+# call train
+trainer.train()
+```
+
+The picture below summarizes the above discussion in more detail. It demonstrates how the design will not contradict internal workings of [`SFTTrainer`].
+- Model is modified and then control passed to [`SFTTrainer`].
+- [`SFTTrainer`] also performs model augmentation internally (e.g., it installs PEFT adapters if `peft_config` is passed in). 
+    * However, [`SFTTrainer`]'s model augmentation should be passed through if configs are omitted (e.g., if `peft_config = None`).
+- [`SFTTrainer`] will prepare model for distributed training (e.g. wrap with `FSDP`) internally. 
+    * thus Plugin implementers need to be aware that `TuningAccelerationPlugin.augmentation` should not interfere with any model preperation that [`SFTTrainer`] will perform.
+
+![Framework](imgs/002-framework.png)
+
+### Acceleration Methods 
+
+A top priority is to incorporate methods that enchance PEFT. While PEFT is known to be memory efficient, it is known to be slower than full-finetuning if not *properly optimized*. Also, another topic of interest is to add support for 4D masks to enable packing while instruction tuning; this acceleration may require some adjustments to the data processing. 
```

Comment:
might help to clarify what we mean by "slower." slower to converge? slower tokens per GPU per second?

Reply:
Thanks for the comment! yes I have now attached an ICLR paper that has a very clear graphic that Full Finetuning actually requires the least tuning time across various PEFT methods.!